File: examples/python-openapi-listener/calculator/get_pet.py
"""
Handler for getPet operation - fetches pet data from petstore API.
Demonstrates using third-party dependencies (requests library) in jackdaw listeners.
"""
import requests
from calculator.types import PetResponse


def handler(event: dict) -> PetResponse:
    """
    Fetch pet information from the petstore API.

    The event contains path parameters from the OpenAPI spec.
    For GET /pet/{petId}, the petId is available in the event.

    Args:
        event: CloudEvent envelope containing the request data

    Returns:
        PetResponse with pet information from the petstore API
    """
    # Extract petId from the event
    # The listener wraps the request in a CloudEvent envelope
    # Path parameters are available in the event data
    pet_id = event.get("petId")

    if not pet_id:
        raise ValueError("petId is required")

    # Call the petstore API
    # Using https://petstore3.swagger.io which is commonly used in CTK tests
    url = f"https://petstore3.swagger.io/api/v3/pet/{pet_id}"

    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status()

        # Parse the JSON response from petstore
        pet_data = response.json()

        # Return the pet data - jackdaw will validate it against PetResponse schema
        result: PetResponse = {
            "id": pet_data.get("id", 0),
            "name": pet_data.get("name", "Unknown"),
            "status": pet_data.get("status", "unknown"),
        }

        # Add optional fields if present
        if "category" in pet_data:
            result["category"] = pet_data["category"]
        if "photoUrls" in pet_data:
            result["photoUrls"] = pet_data["photoUrls"]
        if "tags" in pet_data:
            result["tags"] = pet_data["tags"]

        return result

    except requests.exceptions.HTTPError as e:
        if e.response.status_code == 404:
            raise ValueError(f"Pet {pet_id} not found") from e
        raise ValueError(f"Failed to fetch pet: {e}") from e
    except requests.exceptions.RequestException as e:
        raise ValueError(f"Network error fetching pet: {e}") from e

File: examples/javascript-openapi-listener/src/get_pet.js
/**
 * JavaScript handler for GetPet operation - fetches pet data from petstore API.
 * Demonstrates using npm dependencies in jackdaw listeners.
 *
 * @param {Object} request - Request containing petId
 * @param {number} request.petId - ID of pet to fetch
 * @returns {Promise<Object>} Pet information from the petstore API
 */
// Using axios - a popular HTTP client from npm
// This demonstrates using npm dependencies with Node.js
// Install dependencies with: npm install
import axios from "axios";

export async function handler(request) {
  const petId = request.petId;

  if (!petId) {
    throw new Error("petId is required");
  }

  // Call the petstore API using axios (from npm)
  // Using https://petstore3.swagger.io which is commonly used in tests
  const url = `https://petstore3.swagger.io/api/v3/pet/${petId}`;

  try {
    // axios automatically parses JSON responses
    const response = await axios.get(url);
    const petData = response.data;

    // Return the pet data - jackdaw will validate it against the OpenAPI schema
    const result = {
      id: petData.id || 0,
      name: petData.name || "Unknown",
      status: petData.status || "unknown",
    };

    return result;
  } catch (error) {
    // Check for 404 errors
    if (error.response?.status === 404) {
      throw new Error(`Pet ${petId} not found`);
    }
    if (error.response?.status) {
      throw new Error(`Failed to fetch pet: HTTP ${error.response.status}`);
    }
    // Network or other errors
    throw new Error(`Network error fetching pet: ${error.message}`);
  }
}
